### PR TITLE
Feature/support multiple report actions

### PR DESCRIPTION
**Pull request by wtritch**
I abstracted out a ReportActionService that is responsible for:
1. providing actions for a Report instance
2. performing the provided actions on user interaction

image

### CODE REVIEW

File: webapp/src/main/webapp/src/app/report/report-action.component.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Responsible for providing a UI displaying and performing an action
+ * on a Report.
+ */
+@Component({
+  selector: 'report-action',
+  templateUrl: './report-action.component.html'
+})
+export class ReportActionComponent implements OnInit {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/report/report-action.component.ts]: I'd be tempted to generify this and just say MenuedButton or something. And pass `report` and `report.label` to `@Input() value, text` attr

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/Report.java [hunk 2]
```diff
@@ -169,6 +169,8 @@ public Builder created(final Instant created) {
 
         public Builder metadata(final Map<String, String> metadata) {
             this.metadata.clear();
+            if (metadata == null) return this;
```
Comment (wtritch) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/Report.java]: Small change to fix an error seen during demo.

File: webapp/src/main/webapp/src/app/report/report-action.component.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+<!-- One Action -->
+<button *ngIf="reportActions.length == 1" [title]="report.label" class="btn btn-info btn-xs label-max-width" (click)="performAction(reportActions[0])">
+  <i *ngIf="reportActions[0].icon" class="fa {{ reportActions[0].icon }} mr-xs" aria-hidden="true"></i> {{report.label}}
+</button>
+
+<!-- Multiple Actions -->
+<popup-menu *ngIf="reportActions.length > 1"
+            [item]="report" [actions]="menuActions"
+            [text]="report.label"></popup-menu>
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/report/report-action.component.html]: I broke out a report-action component to handle displaying the report "label" in the My Reports page.  It'll change the display based upon the actions available for the report.

File: webapp/src/main/webapp/src/app/report/report-action.service.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+   * Get the list of actions that can be performed on the given report.
+   *
+   * @param {Report} report     A report
+   * @returns {ReportAction[]}  The actions available for the report
+   */
+  public getActions(report: Report): ReportAction[] {
+    return this.actionProviders
+      .find(provider => provider.accept(report))
+      .getActions(report);
+  }
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/report/report-action.service.ts]: Find the first accepting provider and ask it for the actions for the given report.

File: webapp/src/main/webapp/src/app/report/report-action.service.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+/**
+ * Implementations of this interface are responsible for optionally
+ * accepting a report instance.  If accepted, the implementation is
+ * responsible for providing all actions that can be performed on
+ * the given report.
+ */
+interface ActionProvider {
+  accept(report: Report): boolean;
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/report/report-action.service.ts]: accept -> supports

File: webapp/src/main/webapp/src/app/report/report-action.service.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  public performAction(action: ReportAction): void {
+    switch (action.type) {
+      case ActionType.Download:
+        this.performDownload(action);
+        return;
+      case ActionType.Navigate:
+        this.performNavigate(action);
+        return;
+    }
+  }
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/report/report-action.service.ts]: perform the given action

File: webapp/src/main/webapp/src/app/report/report.service.ts [hunk 7]
```diff
@@ -112,9 +112,6 @@ export class ReportService {
    */
   public getExamReport(reportId: number): Observable<Download> {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/report/report.service.ts]: maybe rename to getReportContent / ReportFile

File: webapp/src/main/webapp/src/app/report/report-action.component.ts (hunk 1 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/app/report/report-action.component.ts]: I'm not sure we can generify this and gain much.  It also has knowledge of the ReportActionService, so we'd have to generify that layer as well...

File: webapp/src/main/webapp/src/app/report/report-action.service.ts (hunk 5 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/app/report/report-action.service.ts]: done

File: webapp/src/main/webapp/src/app/report/report.service.ts (hunk 7 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/app/report/report.service.ts]: I like getReportContent

**Pull request merged**